sandy/core/cov.py

```diff
@@ -474,57 +474,80 @@ def get_corr(self):
+        This method does not require any pre-processing of the covariance
+        data, e.g. removing zeros from the matrix diagonal or truncating
+        eigenvalues.
+        
+        >>> c = sandy.CategoryCov(np.diag(np.array([1, 2, 3])))
+        >>> c.invert()
                     0           1           2
         0 1.00000e+00 0.00000e+00 0.00000e+00
         1 0.00000e+00 5.00000e-01 0.00000e+00
         2 0.00000e+00 0.00000e+00 3.33333e-01
 
-        >>> S = sandy.CategoryCov(np.diag(np.array([0, 2, 3])))
-        >>> S.invert()
-                    0           1           2
-        0 0.00000e+00 0.00000e+00 0.00000e+00
-        1 0.00000e+00 5.00000e-01 0.00000e+00
-        2 0.00000e+00 0.00000e+00 3.33333e-01
+        Test product $A^T A = 1$.
+        >>> c = sandy.CategoryCov([[2, 0.5], [0.5, 2]])
+        >>> np.testing.assert_array_almost_equal(c.data @ c.invert(), np.eye(2))
 
-        >>> S = sandy.CategoryCov(np.diag(np.array([0, 2, 3])))
-        >>> S.invert(rows=1)
-                    0           1           2
-        0 0.00000e+00 0.00000e+00 0.00000e+00
-        1 0.00000e+00 5.00000e-01 0.00000e+00
-        2 0.00000e+00 0.00000e+00 3.33333e-01
+        Test inverse of inverse.
+        >>> c = sandy.CategoryCov(np.diag(np.array([1, 2, 3])))
+        >>> np.testing.assert_array_equal(sandy.CategoryCov(c.invert()).invert(), c.data)
+        
+        Test on real ND covariance. With previous implementation this test failed.
+        >>> c = sandy.get_endf6_file("jeff_33", "xs", 10010).get_errorr(err=1, mt=[102]).get_cov(multigroup=False)
+        >>> cinv = c.invert()
+        >>> np.testing.assert_array_almost_equal(c.data, np.linalg.pinv(cinv, hermitian=True))
+        >>> assert (cinv.index == c.data.index).all()
+        >>> assert (cinv.columns == c.data.columns).all()
+        """
+        return pd.DataFrame(
+            np.linalg.pinv(self.data, hermitian=True),
+            index=self.data.index,
+            columns=self.data.columns,
+        )  # do not return CategoryCov because variance can be negative
+
+    def _double_invert(self):
         """
-        index = self.data.index
-        columns = self.data.columns
-        M_nonzero_idxs, M_reduce = reduce_size(self.data)
-        cov = sps.csc_matrix(M_reduce.values)
-        rows_ = cov.shape[0] if rows is None else rows
-        data = sparse_tables_inv(cov, rows=rows_)
-        M_inv = restore_size(M_nonzero_idxs, data, len(self.data))
-        M_inv = pd.DataFrame(M_inv.values, index=index, columns=columns)
-        return self.__class__(M_inv)
+        Test method get a covariance matrix without negative eigs.
+        BEcause of the properties of pinv, this should be equivalent to
+        reconstructing the matrix as `V @ E @ V`, where `V` are the
+        eigenvectors and `E` are truncated eigenvalues.
+
+        Returns
+        -------
+        `sandy.CategoryCov`
+            matrix without problematic eigenvalues

```

Comment:
It will return the pinv of the inverted matrix, i.e. the original matrix but it can have the negative eigenvalues which are problematic in our case I think

Reply:
I will change it